src/App.tsx

```diff
@@ -1,26 +1,202 @@
+    try {
+      setTodos(prevTodos =>
+        prevTodos.map(todo =>
+          todo.id === todoId ? { ...todo, deleting: 'deleting' } : todo,
+        ),
+      );
+
+      await deleteTodo(todoId);
+
+      setTodos(prevTodos => prevTodos.filter(todo => todo.id !== todoId));
+    } catch {
+      setError('Unable to delete a todo');
+      setTodos(prevTodos =>
+        prevTodos.map(todo =>
+          todo.id === todoId ? { ...todo, deleting: 'idle' } : todo,
+        ),
+      );
+    } finally {
+      inputRef.current?.focus();
+    }
+  };
+
+  const handleClearCompleted = async () => {
+    const completedTodos = todos.filter(todo => todo.completed);
+    const deletePromises = completedTodos.map(todo => deleteTodo(todo.id));
+
+    try {
+      await Promise.all(deletePromises);
+      setTodos(prevTodos => prevTodos.filter(todo => !todo.completed));
+    } catch {
+      setError('Unable to delete completed todos');
+    }
+  };
+
+  const handleToggleAll = () => {
+    const allCompleted = todos.every(todo => todo.completed);
+
+    setTodos(todos.map(todo => ({ ...todo, completed: !allCompleted })));
+  };
+
+  const filteredTodos = todos.filter(todo => {
+    if (filter === 'active') {
+      return !todo.completed;
+    }
+
+    if (filter === 'completed') {
+      return todo.completed;
+    }
+
+    return true;
+  });
 
   return (
-    <section className="section container">
-      <p className="title is-4">
-        Copy all you need from the prev task:
-        <br />
-        <a href="https://github.com/mate-academy/react_todo-app-loading-todos#react-todo-app-load-todos">
-          React Todo App - Load Todos
-        </a>
-      </p>
-
-      <p className="subtitle">Styles are already copied</p>
-    </section>
+    <div className="todoapp">
+      <h1 className="todoapp__title" data-cy="TodoAppTitle">
+        todos
+      </h1>
+      <div className="todoapp__content">
+        <header className="todoapp__header">
+          <button
+            type="button"
+            aria-label="Mark all as completed"
+            className={classNames('todoapp__toggle-all', {
+              active: todos.every(todo => todo.completed),
+            })}
+            onClick={handleToggleAll}
+            data-cy="ToggleAllButton"
+          />
+          <TodoForm
+            onAddTodo={handleAddTodo}
+            newTodoTitle={newTodoTitle}
+            setNewTodoTitle={setNewTodoTitle}
+            isSubmitting={isSubmitting}
+            inputRef={inputRef}
+          />
+        </header>
+
+        <TodoList
+          todos={filteredTodos}
+          loading={loading}
+          setTodos={setTodos}
+          onDeleteTodo={handleDeleteTodo}
+        />
+
+        {tempTodo && (
+          <div className="todo temp-todo" data-cy="TempTodo">
+            <div className="loader"></div>
+            <span className="todo__title">{tempTodo.title}</span>
+          </div>
+        )}
```

Comment:
temp todo should be full todo item with className is-active for TodoLoader.
it can help to pass tests and prevent such view 
![Screenshot 2024-11-23 at 18 44 30](https://github.com/user-attachments/assets/0428f24f-8fb1-49ef-b0a4-477526340d69)

compare your implementation with [working example](https://mate-academy.github.io/react_todo-app-with-api/)